traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
-            and getattr(self, "trait", None) is not None
-        ):
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed, self[:])
-            )
+        self[:] = sorted(self, key=key, reverse=reverse)
+        added = self[:]
+        index = slice(0, len(self), None)
+
+        self.notify(index, removed, added)
 
     def reverse(self):
+        """ Reverse the order of the items in the list *IN PLACE*.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice between 0 and len(self).
+            added : list
+                A copy of the list after reversing the order of items.
+            removed : list.
+                A copy of the list before reversing the order of items.
+
+        """
         removed = self[:]
-        if len(self) > 1:
-            list.reverse(self)
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(0, removed, self[:])
-                )
+        self[:] = self[::-1]
+        added = self[:]
 
-    def pop(self, *args):
-        if not hasattr(self, "trait"):
-            return list.pop(self, *args)
-        if self.trait.minlen < len(self):
-            if len(args) > 0:
-                index = args[0]
-            else:
-                index = -1
+        index = slice(0, len(self), None)
 
-            try:
-                removed = [self[index]]
-            except Exception:
-                pass
+        self.notify(index, removed, added)
 
-            result = list.pop(self, *args)
+    def __iadd__(self, other):
+        """ Implements the self += value operator.
+
+        Parameters
+        ----------
+        other : Any
+            The item to add to the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
 
-            if self.name_items is not None:
-                if index < 0:
-                    index = len(self) + index + 1
+        """
+        self.extend(other)
+        return self
 
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed)
-                )
+    def __imul__(self, count):
+        """ Implements the self *= value operator.
+
+        Parameters
+        ----------
+        count : int
+            Number of times to duplicate the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
+
+        """
+        if isinstance(count, int):
```

Comment:
This `isinstance` check doesn't match the behaviour of Python (where any integer-like is permitted). If possible, we should avoid the instance check altogether and use the type validation implicit in `self * count`.

Reply:
Removed it.